src/DropdownBase.js

```diff
@@ -0,0 +1,307 @@
+
+const TOGGLE_REF = 'toggle-btn';
+
+export default class DropdownBase extends React.Component {
+  constructor(props, Toggle=DropdownToggle) {
+    super(props);
+
+    this.Toggle = Toggle;
+
+    this.toggleOpen = this.toggleOpen.bind(this);
+    this.handleClick = this.handleClick.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.handleRequestClose = this.handleRequestClose.bind(this);
+    this.handleSelect = this.handleSelect.bind(this);
+    this.extractChildren = this.extractChildren.bind(this);
+
+    this.refineMenu = this.refineMenu.bind(this);
+    this.refineToggle = this.refineToggle.bind(this);
+
+    this.childExtractors = [{
+      key: 'toggle',
+      matches: child => child.type === this.Toggle,
+      refine: this.refineToggle
+    }, {
+      key: 'menu',
+      matches: child => child.type === DropdownMenu || child.type !== MenuItem,
+      refine: this.refineMenu
+    }];
+
+    this.state = {
+      open: false
+    };
+  }
+
+  toggleOpen() {
+    let open = !this.state.open;
+    this.setState({open});
+  }
+
+  handleClick(event) {
+    if (event.defaultPrevented) {
+      return;
+    }
+
+    this.toggleOpen();
+  }
+
+  handleKeyDown(event) {
+    switch(event.keyCode) {
+      case keycode.codes.down:
+        if (!this.state.open) {
+          this.toggleOpen();
+        }
+        if (this.refs.menu.focusNext) {
+          this.refs.menu.focusNext();
+        }
+        event.preventDefault();
+        break;
+      case keycode.codes.esc:
+      case keycode.codes.tab:
+        if (this.state.open) {
+          this.handleRequestClose(event);
+        }
+        break;
+    }
+  }
+
+  handleRequestClose(event) {
+    if (!this.state.open) {
+      return;
+    }
+
+    this.toggleOpen();
+
+    if (event && event.type === 'keydown' && event.keyCode === keycode.codes.esc) {
+      event.preventDefault();
+      event.stopPropagation();
+      React.findDOMNode(this.refs[TOGGLE_REF]).focus();
+    }
+  }
+
+  handleSelect(event, selectEvent) {
+    if (selectEvent.isSelectionPrevented()) {
+      return;
+    }
+
+    this.handleRequestClose(event);
+  }
+
+  extractChildren() {
+    let open = this.props.open !== undefined ? this.props.open : this.state.open;
+    let children = [];
+    let set = {};
+
+    React.Children.forEach(this.props.children, child => {
+      let extractor = find(this.childExtractors, x => x.matches(child));
+
+      if (extractor) {
+        set[extractor.key] = set[extractor.key] || child;
+      } else {
+        children.push(child);
+      }
+    });
+
+    this.childExtractors.forEach(extractor => {
+      set[extractor.key] = extractor.refine(set[extractor.key], children, open);
+    });
+
+    return {
+      open,
+      ...set
+    };
+  }
+
+  refineMenu(menu, children, open) {
+    const menuProps = {
+      ref: 'menu',
+      open,
+      labelledBy: this.props.id
+    };
+
+    if (menu === undefined) {
```

Comment:
cc @jquense 

This is actually the opposite of what we decided to do for `Modal`, i.e. the support for custom modal dialogs is going into react-overlays, while the R-B `Modal` intentionally has fairly limited support (and a different API) for specifying a custom modal dialog class.

It might be nice to be consistent here if possible.